pkg/obfuscator/tracker.go

```diff
@@ -0,0 +1,84 @@
+package obfuscator
+
+import (
+	"sync"
+
+	"k8s.io/klog/v2"
+)
+
+type GenerateReplacement func(string) string
+
+// ReplacementTracker is used to track and generate replacements used by obfuscators
+type ReplacementTracker interface {
+	// Report returns a mapping of strings which were replaced.
+	Report() map[string]string
+
+	// AddReplacement will add a replacement along with its original string to the report.
+	// If there is an existing value that does not match the given replacement, it will exit with a non-zero status.
+	AddReplacement(original string, replacement string)
+
+	// GetReplacement returns the previously used replacement if already set. If the replacement is not present then it
+	// uses the GenerateReplacement function to generate a replacement. If the generator is nil then it returns an empty
+	// string. The original parameter must be used for lookup and the key parameter to generate the replacement.
+	GetReplacement(original string, key string, generator GenerateReplacement) string
```

Comment:
do you mind if we stick to the java nomenclature of PutIfAbsent?

Reply:
The Java name is such because there a a ton of functions like `Get`, `GetIfAbsent`, `GetIfPresent` and so on. But in this case there is only one function so `GetReplacement` is appropriate IMO.

Reply:
maybe GenerateIfAbsent? It certainly does more than just getting something

Reply:
But it also does a simple `Get` if the generator is `nil`.

Reply:
shall we split it into two funcs then, one just doing a get and one with a generator `GenerateIfAbsent` maybe?

Reply:
Why?

Reply:
Separation of concerns I'd say. One is just doing a Get that returns empty when there is nothing, otherwise the result. The other is atomically replacing a key if there is no result and returns that. It's two different things that we shouldn't really mix IMHO.

Reply:
Renamed.

Reply:
thanks, apologies for bikeshedding